src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala

```diff
@@ -0,0 +1,447 @@
+    case RequestFailed(peer, reason) if state.isBodiesFetcher(sender()) =>
+      log.debug(s"Failed bodies request from peer {} with reason {}", peer, reason)
+      blacklistIfHandshaked(peer, reason)
+      state.withoutBodiesFetcher() |> fetchBlocks
+  }
+
+  private def handleStateNodeMessages(state: FetcherState): Receive = {
+    case FetchStateNode(hash) => fetchStateNode(hash, sender(), state)
+    case RetryFetchStateNode(hash, replyTo) => fetchStateNode(hash, replyTo, state)
+    case ResponseReceived(peer, NodeData(values), _) if state.isStateNodeFetcher(sender()) =>
+      log.debug("Received state node response from peer {}", peer)
+      state.stateNodeFetcher.foreach(fetcher => {
+        val validatedNode = values
+          .asRight[String]
+          .ensure(s"Empty response from peer $peer, blacklisting")(_.nonEmpty)
+          .ensure(s"Fetched node state hash doesn't match requested one, blacklisting peer")(nodes =>
+            fetcher.hash == kec256(nodes.head))
+
+        validatedNode match {
+          case Left(err) =>
+            log.debug(err)
+            blacklist(peer.id, syncConfig.blacklistDuration, err)
+            fetchStateNode(fetcher.hash, fetcher.replyTo, state)
+          case Right(node) =>
+            fetcher.replyTo ! FetchedStateNode(NodeData(node))
+            context become started(state.withoutStateNodeFetcher())
+        }
+      })
+    case RequestFailed(peer, reason) if state.isStateNodeFetcher(sender()) =>
+      log.debug(s"Failed state node request from peer {} with reason {}", peer, reason)
+      blacklistIfHandshaked(peer, reason)
+      state.stateNodeFetcher.foreach(fetcher => {
+        fetchStateNode(fetcher.hash, fetcher.replyTo, state.withoutHeadersFetcher())
+      })
+  }
+
+  private def handleNewBlockMessages(state: FetcherState): Receive = {
+    case MessageFromPeer(NewBlockHashes(_), _) =>
+      val newState = state.notFetchedTopHeader()
+      fetchBlocks(newState)
+    case MessageFromPeer(NewBlock(block, _), peerId) if state.isOnTop =>
+      val lastReadyBlock = state.lastReadyBlockNumber
+      if (Block.number(block) == lastReadyBlock + 1) {
+        state.importer ! ImportNewBlock(block, peerId)
+        val newState =
+          peerById(peerId).fold(state)(peer => state.withPeerForBlocks(peer, Seq(Block.number(block))))
+        context become started(newState)
+      } else {
+        val newState = state.notFetchedTopHeader()
+        fetchBlocks(newState)
+      }
+  }
+
+  private def fetchBlocks(state: FetcherState): Unit = {
+    val newState = state |> tryFetchHeaders |> tryFetchBodies
+
+    if (!state.isFetching && !newState.hasFetchedTopHeader) {
+      scheduler.scheduleOnce(10.seconds, context.self, FetchBlocks)
+    }
+
+    context become started(newState)
+  }
+
+  private def tryFetchHeaders(fetcherState: FetcherState): FetcherState =
+    Some(fetcherState)
+      .filterNot(_.isFetchingHeaders)
```

Comment:
tbh i do not like this filter trains, on each one of them i must consider hard what filter does, does it leave elements which pass the predicate or other way around.

Reply:
Switching it into plain `filter` would be a good fix to you? TBH I had the same problem at the beginning but now it's pretty clear to me. Also it seems nicer to me to write (and read) `filterNot(sth)` instead of `filter(!sth)`

Reply:
yep simple filter would be good 👍